test/core/authorship/proposer_test.cpp

```diff
@@ -0,0 +1,158 @@
+  Block expected_block{{}, {{{5, 4, 3}}}};
+};
+
+/**
+ * @given BlockBuilderApi creating inherent extrinsics @and TransactionPool
+ * returning extrinsics
+ * @when Proposer created from these BlockBuilderApi and TransactionPool is
+ * trying to create block
+ * @then Block is created and it is equal to the block baked in BlockBuilder
+ */
+TEST_F(ProposerTest, CreateBlockSuccess) {
+  // given
+  // we push 2 xts: 1 from inherent_xts and one from transaction pool
+  EXPECT_CALL(*block_builder_, pushExtrinsic(_))
+      .WillOnce(Return(outcome::success()))
+      .WillOnce(Return(outcome::success()));
+
+  // getReadyTransaction will return vector with single transaction
+  EXPECT_CALL(*transaction_pool_, getReadyTransactions())
+      .WillOnce(Return(std::vector<Transaction>{{}}));
+
+  EXPECT_CALL(*block_builder_, bake()).WillOnce(Return(expected_block));
+
+  // when
+  auto block_res =
+      proposer_.propose(expected_block_id_, inherent_data_, inherent_digest_,
+                        kagome::time::now() + 10000000  // big enough deadline
+      );
+
+  // then
+  ASSERT_TRUE(block_res);
+  ASSERT_EQ(expected_block, block_res.value());
+}
+
+/**
+ * @given BlockBuilderApi creating inherent extrinsics @and TransactionPool
+ * returning extrinsics @and BlockBuilder that cannot accept extrinsics
+ * @when Proposer created from these BlockBuilderApi and TransactionPool is
+ * trying to create block
+ * @then Block is not created
+ */
+TEST_F(ProposerTest, CreateBlockFailsWhenXtNotPushed) {
+  // given
+  EXPECT_CALL(*block_builder_, pushExtrinsic(inherent_xts[0]))
+      .WillOnce(Return(outcome::failure(boost::system::error_code{})));
+
+  // when
+  auto block_res =
+      proposer_.propose(expected_block_id_, inherent_data_, inherent_digest_,
+                        kagome::time::now() + 10000000  // big enough deadline
+      );
+
+  // then
+  ASSERT_FALSE(block_res);
+}
+
+/**
+ * @given BlockBuilderApi creating inherent extrinsics @and TransactionPool
+ * returning extrinsics
+ * @when Proposer created from these BlockBuilderApi and TransactionPool is
+ * trying to create block @and deadline is passed
+ * @then Block is created @but extrinsics from transaction pool are not pushed
+ */
+TEST_F(ProposerTest, CreateBlockFailsWhenDeadlinePassed) {
+  // given
+  // we push 1 xt from inherent_xts. Xt from transaction pool will not be pushed
+  // as deadline already passed
+  EXPECT_CALL(*block_builder_, pushExtrinsic(_))
+      .WillOnce(Return(outcome::success()));
+
+  EXPECT_CALL(*block_builder_, bake()).WillOnce(Return(expected_block));
+
+  // getReadyTransaction will return vector with single transaction
+  EXPECT_CALL(*transaction_pool_, getReadyTransactions())
+      .WillOnce(Return(std::vector<Transaction>{{}}));
+
+  // when
+  auto block_res =
+      proposer_.propose(expected_block_id_, inherent_data_, inherent_digest_,
+                        kagome::time::now() - 20);
```

Comment:
What is 20? I would add a `kagome::time::time_t before()` function and use `kagome::time::before(std::chrono::milliseconds(20)` or `seconds` or something else.

Reply:
We need to test the behaviour that when deadline passed no extrinsic is added. That's why I added -20. This is the only place when we do this, so no need for a separate function for now

Reply:
ok if for 1 test only